<a href="https://colab.research.google.com/github/QuirozAlaniz/Datos_Temporales/blob/main/Actividad_Calificada_07_Datos_Temporales_269723.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformación de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | 26 de Septiembre de 2026 |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |Francisco Daniel Quiroz Alaniz|
| **Matrícula** |269723|

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesión: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripción |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzó el **domingo 10 de marzo**: a las 02:00 el reloj saltó a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversión y validación** | 20 | 18–20: conversión con formato explícito y verificación de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusión. | 1–13: conversión sin verificar o validaciones incompletas; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duración y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusión argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entregó o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretación fundamentada. | 18–22: componentes y agrupaciones correctos; interpretación incompleta. | 1–17: agrupaciones con errores o sin interpretación. | 0: no se entregó o es incorrecto. |
| **4. Dataset limpio y comparación** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparación entre periodos con filtros temporales correcta; registro de decisión completo. | 14–17: dataset limpio correcto; comparación o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificación. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [795]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversión y validación (20 puntos)

### 1.1 Conversión (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversión no generó `NaT`: compara el número de faltantes antes y después.

In [796]:
# Escribe tu código aquí
#1. Usamos el atributo .dtypes para obtener el tipo de dato de ambas columnas
print(f'Los tipos de datos de pickup y dropoff son: \n  {taxis[["pickup", "dropoff"]].dtypes}')   #pickup es la fecha y hora en que inicia el viaje, dropoff cuando termina

Los tipos de datos de pickup y dropoff son: 
  pickup     object
dropoff    object
dtype: object


In [797]:
#Vemos cuántos valores faltan.
print("El número de faltantes es:\n", taxis[["pickup", "dropoff"]].isna().sum() )    #No hay ningún faltante


El número de faltantes es:
 pickup     0
dropoff    0
dtype: int64


In [798]:
#2. Convertimos ambas columnas a tipo fecha usando el método indicado.
taxis["pickup"] = pd.to_datetime(taxis["pickup"], format = '%Y-%m-%d %H:%M:%S')
taxis["dropoff"] = pd.to_datetime(taxis["dropoff"], format= '%Y-%m-%d %H:%M:%S')

In [799]:
print( f'Los valores NaT son:\n {taxis[["pickup", "dropoff"]].isnull().sum()}')  #No hay NaT
print()
print(f'Podemos ver que no se generaron NaT')

Los valores NaT son:
 pickup     0
dropoff    0
dtype: int64

Podemos ver que no se generaron NaT


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [800]:
# Escribe tu código aquí
#1. Usamos .min() y .max() para obtener las fechas mínima y máxima
minima_pickup = taxis["pickup"].min()
maxima_pickup = taxis["pickup"].max()
minima_dropoff = taxis["dropoff"].min()
maxima_dropoff = taxis["dropoff"].max()
print(f'pickup: \n minima = {minima_pickup} \n maxima = {maxima_pickup}')
print()
print(f'dropoff: \n minima = {minima_dropoff} \n maxima = {maxima_dropoff}')
print()
print("Podemos ver que sí hay registros que no están en marzo solamente")

pickup: 
 minima = 2019-02-28 23:29:03 
 maxima = 2019-03-31 23:43:45

dropoff: 
 minima = 2019-02-28 23:32:35 
 maxima = 2019-04-01 00:13:58

Podemos ver que sí hay registros que no están en marzo solamente


In [801]:
#2. Para contar los viajes cuyo pickup no está en marzo de 2019 usamos el atributo dt.month para extraer el número de mes y lo comparamos con el =! 3
viajes_nomarzo =  taxis[taxis["pickup"].dt.month != 3]
print(f'El número de viajes cuyo pickup no está en marzo es: {len(viajes_nomarzo)} \ny es el siguiente:\n {viajes_nomarzo}')
#Nos genera un registro cuyo pickup está en Febrero.

El número de viajes cuyo pickup no está en marzo es: 1 
y es el siguiente:
                   pickup             dropoff  passengers  distance  fare  tip  \
6203 2019-02-28 23:29:03 2019-02-28 23:32:35           1       0.9   5.0  0.0   

      tolls  total  color payment  pickup_zone                   dropoff_zone  \
6203    0.0    6.3  green    cash  Old Astoria  Long Island City/Queens Plaza   

     pickup_borough dropoff_borough  
6203         Queens          Queens  


In [802]:
#3. #Usamos también el atributo dt.month
pickup_dropoff = taxis[(taxis["pickup"].dt.month == 3) &  (taxis["dropoff"].dt.month != 3) ]
print(f'El número de registros cuyo pickup es de marzo y el dropoff no es de marzo es: {len(pickup_dropoff)} \n y es el siguiente: \n {pickup_dropoff}')

El número de registros cuyo pickup es de marzo y el dropoff no es de marzo es: 1 
 y es el siguiente: 
                  pickup             dropoff  passengers  distance  fare  tip  \
591 2019-03-31 23:43:45 2019-04-01 00:13:58           5     12.25  37.0  0.0   

     tolls  total   color payment        pickup_zone  \
591    0.0   40.8  yellow    cash  LaGuardia Airport   

                  dropoff_zone pickup_borough dropoff_borough  
591  Williamsburg (South Side)         Queens        Brooklyn  


### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [803]:
# Escribe tu código aquí
#1. Para filtrar los viajes el 10 de marzo usaré dt.hour.
filtro_10 = taxis[taxis["pickup"].dt.day == 10]                          #Filtramos por pickup que son del día 10
conteo_hora = filtro_10["pickup"].dt.hour.value_counts().sort_index()     #accedemos a horas de columna pickup y contamos ocurrencias con value_counts y ordenamos los índices
print(f'Las horas registradas el 10 de marzo son: \n {conteo_hora} ')    #Contamos cuántas filas tiene el filtro, no tiene ninguna.
print()
print(f'Podemos observar que no hay registros de la hora 2')



Las horas registradas el 10 de marzo son: 
 pickup
0     11
1      7
3      6
4      7
5      1
6      2
7      5
8      7
9      9
10    13
11    11
12    17
13     8
14    13
15     9
16    10
17    14
18     7
19     3
20     4
21     3
22    10
23     8
Name: count, dtype: int64 

Podemos observar que no hay registros de la hora 2


In [804]:
#2
pickup_zona = taxis["pickup"].dt.tz_localize('America/New_York', nonexistent = "NaT") #Etiquetamos los datos con la región especificada y los inválidos los hacemos NaT
total_nat = pickup_zona.isna().sum()
print(f'El número de NaT es: {total_nat}')

El número de NaT es: 0


### 1.4 Conclusión (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.

**Respuesta 1.4:**

_Escribe aquí tu respuesta._
\
Un problema temporal que encontré es que hay un registro en pickup que es de Febrero y también hay un registro que tiene pickup de marzo y dropoff de Abril.
\
El primer caso no debería ser un registro porque tanto el pickup como el dropoff ocurrieron en febrero, sin embargo el segundo caso podría llegar a ser un registro válido porque una parte del viaje (el dropoff) ocurrió en marzo, aunque la regla que elegiría sería tomar como Marzo aquellos cuyo dropoff sea Marzo.
\
También encontré que no existe en el 10 de marzo una hora correspondiente a las 2:00 porque esa hora no existió por ser el cambio de horario, por eso al usar .dt.tz_localize con argumento America/New_York no arrojó ningún NaT, porque era lógico que no habría ningún registro en esa hora.

---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duración de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duración en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [805]:
# Escribe tu código aquí
#1. Creamos la columna duración
taxis["duracion"] =  taxis["dropoff"] - taxis["pickup"]

In [806]:
#2, Creamos la columna duracion_min
taxis["duracion_min"] = taxis["duracion"].dt.total_seconds()/60   #Como dt.total_seconds nos da el total de segundos, hay que dividirlo entre 60 para que nos de los minutos.

In [807]:
#3. Resumen estadístico.
resumen = taxis["duracion_min"].describe()  #usamos el método .describe() para que nos muestre un resumen estadístico
print(resumen)
print()
print("Se ve que hay registros con duración de 0 minutos y quizá outliers porque el máximo es de 107 minutos cuando el IQR es prácticamente 12")

count    6433.000000
mean       14.349617
std        11.643892
min         0.000000
25%         6.500000
50%        10.900000
75%        18.516667
max       107.666667
Name: duracion_min, dtype: float64

Se ve que hay registros con duración de 0 minutos y quizá outliers porque el máximo es de 107 minutos cuando el IQR es prácticamente 12


### 2.2 Duraciones anómalas (6 pts)

Cuenta los viajes con:

- Duración **negativa** (termina antes de iniciar).
- Duración **cero**.
- Duración **menor a 1 minuto** (incluye los de duración cero).
- Duración **mayor a 3 horas**.

Muestra los viajes de duración cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [808]:
# Escribe tu código aquí
duracion_negativa = taxis [taxis["duracion_min"] < 0 ]
duracion_cero = taxis [taxis["duracion_min"] == 0]
duracion_menormin = taxis[taxis["duracion_min"] < 1 ]
duracion_mayor3hrs = taxis[taxis["duracion_min"] > 180]
print(f'El número de viajes con duración menor a 0 es: {len(duracion_negativa)}')
print()
print(f'El número de viajes con duración cero es: {len(duracion_cero)}')
print()
print(f'El número de viajes conduración menor a un minuto es: {len(duracion_menormin)}')
print()
print(f'El número de viajes con duración mayor a tres horas es: {len(duracion_mayor3hrs)}')


El número de viajes con duración menor a 0 es: 0

El número de viajes con duración cero es: 6

El número de viajes conduración menor a un minuto es: 72

El número de viajes con duración mayor a tres horas es: 0


### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duración mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condición)`, que deja `NaN` donde la condición es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [809]:
# Escribe tu código aquí
#1
taxis["velocidad_mph"] = taxis["distance"]/(taxis["duracion_min"]/60)
taxis["velocidad_mph"] = taxis["velocidad_mph"].where(taxis["duracion_min"] > 0)

In [810]:
#2
print(f'En total hay: {len ( taxis[taxis["velocidad_mph"] > 60 ] )} registros que tienen velocidad mayor a 60 mph')

En total hay: 11 registros que tienen velocidad mayor a 60 mph


In [811]:
#3
print(f'Los cinco registros que van a mayor velocidad son:\n {taxis[["pickup", "dropoff", "duracion_min", "distance", "velocidad_mph", "fare"]].sort_values(by = "velocidad_mph", ascending = False).head(5)}' )

Los cinco registros que van a mayor velocidad son:
                   pickup             dropoff  duracion_min  distance  \
770  2019-03-02 03:16:59 2019-03-02 03:17:06      0.116667      9.40   
6089 2019-03-17 23:16:34 2019-03-17 23:16:52      0.300000     10.10   
491  2019-03-07 07:11:33 2019-03-07 07:11:39      0.100000      1.60   
5284 2019-03-22 06:24:02 2019-03-22 06:24:11      0.150000      1.70   
3241 2019-03-24 17:19:27 2019-03-24 17:19:29      0.033333      0.34   

      velocidad_mph  fare  
770     4834.285714   2.5  
6089    2020.000000  12.0  
491      960.000000   2.5  
5284     680.000000   2.5  
3241     612.000000  65.0  


### 2.4 Criterio de exclusión (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

**Respuesta 2.4:**

_Escribe aquí tu respuesta._
\
Los viajes que excluiría son aquellos cuya velocidad no es lógica, de hecho hay uno con velocidad de 4834 millas por hora con una duración de 0.11 minutos y otro de 2020 mph con una duración de 0.3 minutos, además otros que rebasan los 600 mph. El criterio podría ser eliminar los que rebasen los 100 millas por hora.
\
\
a) De acuerdo con 1.2, hay un registro cuyo pickup y dropoff pertenece a Febrero, entonces no debería estar en el dataset. Además hay uno que comienza en Marzo pero termina en Abril. Para esta casuística lo mejor es establecer una regla: todos los registros van a pertenecer al mes cuyo dropoff suceda en el mes. Entonces estos dos registros de los que hablo serían eliminados.
\
\
b) Hay 72 registros cuya duración es menor a un minuto. Revisando las columnas de esas filas me encontré que hay registros que duran menos de un minuto pero que realizaron viajes muy cortos, quizá por prisas dentro del aeropuerto de una terminal a otra.
\
\
Por otro lado, de esos 72, hay 6 que duraron exactamente 0 minutos que sí tiene tarifa. Esos deben ser errores de registro, porque no puedes cobrar una tarifa de un viaje que simplemente no se movió. Por eso los consideraría errores de registro.

---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [812]:
# Escribe tu código aquí
taxis["hora"] = taxis["pickup"].dt.hour
taxis["dia_num"] = taxis["pickup"].dt.dayofweek
taxis["dia_nombre"] = taxis["pickup"].dt.day_name()
taxis["fin_semana"] = (taxis["dia_nombre"]=="Saturday") | (taxis["dia_nombre"]=="Sunday")
taxis["fecha"] = taxis["pickup"].dt.date
taxis["cruza_medianoche"] = taxis["pickup"].dt.day != taxis["dropoff"].dt.day

In [813]:
print(f'El número de registros que cruzan la media noche es:\n  {len(taxis[taxis["cruza_medianoche"] == True])}')

El número de registros que cruzan la media noche es:
  50


### 3.2 Patrón por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duración tiene valores extremos.

In [814]:
# Escribe tu código aquí
agrupacion = taxis.groupby("hora").agg(numero_viajes = ("duracion_min","count"), mediana_duracion_min = ("duracion_min", "median")) #Agrupamos por hora y creamos columnas
                                                                                                                                    #con funciones de agregación count y median

In [815]:
maximo = agrupacion["numero_viajes"].idxmax()
minimo = agrupacion["numero_viajes"].idxmin()
print(f'La hora con más viajes es: {maximo} con {agrupacion["numero_viajes"].max()} viajes')
print()
print(f'La hora con menos viajes es: {minimo} con {agrupacion["numero_viajes"].min()} viajes')


La hora con más viajes es: 18 con 417 viajes

La hora con menos viajes es: 5 con 51 viajes


In [816]:
#Código no solicitado pero para responder a la pregunta 3.4
agrupacion2 = taxis.groupby("hora").agg(mediana_duracion_trayecto = ("duracion_min", "median"))
print(agrupacion2.sort_values(by = "mediana_duracion_trayecto", ascending = False).head(3))

      mediana_duracion_trayecto
hora                           
12                    12.775000
15                    12.058333
11                    12.033333


### 3.3 Patrón por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [817]:
# Escribe tu código aquí
tabla = taxis.groupby(["dia_num", "dia_nombre"]).agg(numero_viajes = ("duracion_min", "count"), mediana_duracion_min = ("duracion_min", "median"),
                                                     mediana_fare = ("fare", "median"),  numero_dias = ("fecha", "nunique"))
tabla["promedio_viajes_diasemana"] = tabla["numero_viajes"]/tabla["numero_dias"]


In [818]:
print(f'El patrón por día de la semana es el siguiente:\n {tabla}')

El patrón por día de la semana es el siguiente:
                     numero_viajes  mediana_duracion_min  mediana_fare  \
dia_num dia_nombre                                                      
0       Monday                708             10.541667           9.5   
1       Tuesday               825             11.216667           9.5   
2       Wednesday             966             11.291667           9.5   
3       Thursday              905             11.766667           9.5   
4       Friday               1115             11.150000           9.5   
5       Saturday             1046             10.466667           9.0   
6       Sunday                868              9.983333           9.0   

                    numero_dias  promedio_viajes_diasemana  
dia_num dia_nombre                                          
0       Monday                4                     177.00  
1       Tuesday               4                     206.25  
2       Wednesday             4                  

### 3.4 Interpretación (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duración mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

_Escribe aquí tu respuesta._
\
a) Lo que resalta es que prácticamente la duración de los viajes van de entre 10 y 12 minutos.
\
El precio por vaje se encuentra en 9.5 dólares.
\
Los días de la semana en que se hacen menos viajes son el domingo y el lunes, mientras que en el que se hacen más viajes es el miércoles.
\
Resalta que la hora con más viajes es las 18 horas con 417 viajes, pero la hora con mayor duración mediana es las 12 horas con una duración mediana de 12.7 minutos por trayecto.
\
Una explicación de esto es que a las 6 hay más servicios de taxis por los horarios de entrada a oficinas, mientras que a las 12 quizás los trayectos son más largos.
\
\
b) Fue necesario calcular el promedio porque en el mes hay diferente número de miércoles jueves, viernes, etc, en algunos hay 4 y en otros hay 5. Por eso en términos absolutos hay unos que tienen muchos más viajes que otros, mientras que al comparar los promedios hacemos un efecto de aplanado y los números ya son comparables.  

---
## Actividad 4. Dataset limpio y comparación (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condición y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [819]:
# Escribe tu código aquí
taxis_limpio = taxis

In [820]:
#Eliminamos aquéllos que tienen duración en minutos menor a 0.5
print("El número de filas originalmente es:", len(taxis_limpio))
print()
taxis_limpio = taxis_limpio[ taxis_limpio["velocidad_mph"] < 100]
print("El número de filas después de quitarle aquéllos registros cuya velocidad en millas por hora rebasa los 100 es:", len(taxis_limpio) )
print()
taxis_limpio = taxis_limpio[taxis_limpio["duracion_min"] > 1]
print("El número de filas después de quitarle aquéllos registros cuya duración no pasa del minuto es:", len(taxis_limpio))
print()
taxis_limpio = taxis_limpio[taxis_limpio["dropoff"].dt.month == 3]   #Elegimos solo aquellos que tengan mes de término marzo
print("El número de filas después de quitarle aquéllos registros cuyo servicio no termina en Marzo es:", len(taxis_limpio))
print()

El número de filas originalmente es: 6433

El número de filas después de quitarle aquéllos registros cuya velocidad en millas por hora rebasa los 100 es: 6416

El número de filas después de quitarle aquéllos registros cuya duración no pasa del minuto es: 6358

El número de filas después de quitarle aquéllos registros cuyo servicio no termina en Marzo es: 6356



### 4.2 Comparación entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [821]:
# Escribe tu código aquí
comparacion = taxis_limpio[(taxis_limpio["pickup"].dt.hour >= 7) & (taxis_limpio["pickup"].dt.hour <10) ].groupby("fin_semana").agg(numero_viajes = ("pickup", "count"),
                                                                                                                      mediana_duracion = ("duracion_min", "median"),
                                                                                                                      mediana_velocidad = ("velocidad_mph", "median"))
comparacion.index = ["dia_habil", "fin_de_semana"]                         #Renombramos los índices para que sea más claro.
print(comparacion)

               numero_viajes  mediana_duracion  mediana_velocidad
dia_habil                683         12.033333           9.097978
fin_de_semana            161          7.766667          13.473054


### 4.3 Registro de decisión (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversión**: formato utilizado y verificación.
- **Zona horaria**: en qué zona están los datos y cómo lo verificaste.
- **Criterio de exclusión**: condiciones, umbrales y justificación.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Respuesta 4.3:**

_Escribe aquí tu respuesta._
\
En el presente dataset se realizaron las siguientes acciones:

1. Exploración inicial de tipos arrojando:
\
  a) 6433 filas y 14 columnas
  \
  b) Dos columnas de fecha en formato object

2. Se han convertido las dos columnas:"pickup" y "dropoff" a fecha con formato "Año-Mes-Dia"

3. Los datos están en zona horaria "America/Nueva_York" verificado con "tz_localize()" y "nonexistent = NaT"

4. Se realizó una depuración de filas utilizando tres criterios:
   a) La fecha de "dropoff" debe ser en el mes de marzo como regla general y evitar ambigüedades cuando las fechas de "pickup" y "dropof" en el mismo viaje son diferentes.
   \
   b) La duración del viaje debe ser mayor a un minuto porque los menores son muy probablemente errores de registro.
   \
   c) La velocidad en millas por hora no debe rebasar los 100 porque se trata de una ciudad.
\

5. El resultado es que se han reducido de la siguiente manera:
\
   a) Por quitarle los que rebasan los 100 mph de velocidad: -17
   \
   b) Por quitarle los registros que no pasan del minuto: -58
   \
   c) Por quitarle los registros que no terminan en Marzo: -2
   \
   d) Total de viajes excluidos: -77, filas del dataset limpio: 6356
   
6. Al estudiar el número de viajes entre las 7 y las 9:59 horas, realizados en día hábil versus el fin de semana, se encontró que:
\
   a)  el número de viajes hechos en día hábil supera por mucho al número de viajes realizados el fin de semana: 683 VS 161 respectivamente
   \
   b) La mediana de duración de los viajes en día hábil es de 12.03 minutos mientras que los realizados en fin de semana son de 7.7, es decir 5 minutos más cortos.
   \
  c) La mediana de la velocidad de los viajes en día hábil es de 9.09 mph, mientras que la mediana de la velocidad de los realizados el fin de semana es de 13.4 mph, es decir, duran menos pero porque van más rápido.
  \
  Esto se debe muy probablemente a que en día hábil las carreteras y calles de la ciudad de Nueva York se encuentran saturadas con gente que va al trabajo o a la escuela a esa hora.

---
## Reflexión final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situación la inferencia habría producido errores silenciosos?
2. Supón que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Cómo lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duración de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**

_Escribe aquí tu respuesta._
\
1. Darle el formato explícito a pandas en vez de dejar que lo infiera permitió asegurar que las extracciones de mes para hacer la depuración fuera exacta. También facilitó la producción de estadísticas por día de la semana.
\
Una inferencia habría producido errores en el caso en que hay muchos formatos diferentes en los registros de fechas.
\
2. El horario local de Nueva York está 4 y 5 horas por detrás del horario UTC. Son dos diferencias porque en marzo el horario cambia, provocando una hora adicional de diferencia.
\
Con esto en cuenta, dropoff siempre estaría detrás del horario de pickup, haciendo duraciones de trayecto negativas, lo cual no tendría sentido.
\
Lo habría detectado precisamente por las duraciones negativas que me arrojaría.
\
3. Si se entrenara un modelo para predecir la duración de un viaje con el dataset original y usando las variables predictoras las demás columnas, las anomalías temporales provocarían un sesgo sobre la duración del viaje porque el modelo aprendería que hay viajes que pueden durar 0 minutos y podría arrojar precisamente del punto A al punto B una duración de 0 minutos, lo cual es ilógico.


---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.